onnx/defs/math/defs.cc

```diff
@@ -1665,4 +1666,130 @@ ONNX_OPERATOR_SET_SCHEMA(
           }
         }));
 
+void einsumRankInference(
+    ONNX_NAMESPACE::InferenceContext& ctx, std::string equation) {
+
+  const size_t numInputs = ctx.getNumInputs();
+  if (numInputs < 1 || !hasNInputShapes(ctx, static_cast<int>(numInputs))) {
+    return;
+  }
+
+  auto* output_shape = getOutputShape(ctx, 0);
+  std::string  left_equation;
+
+  equation.erase(std::remove(equation.begin(), equation.end(), ' '), equation.end()); // Remove whitespaces
+  auto mid_index = equation.find("->");
+  if (mid_index != std::string::npos) {
+    // Separate right and left hand sides of the equation
+    left_equation = equation.substr(0, mid_index);
+  } else {
+    // No right hand side
+    left_equation = equation;
+  }
+
+  std::string term;
+  size_t num_operands = 0;
+  size_t num_ellipsis = 0;
+  size_t num_ellipsis_indices = 0;
+
+  // Parse the left-hand side
+  std::stringstream str(left_equation);
+  while(std::getline(str, term, ',')) {
+    auto ellipsis_index = term.find("...");
```

Comment:
A check to handle the case where number of operands here exceeds numInputs would be helpful.

Reply:
This is in line 1714

Reply:
The problem is what happens in line 1701 with "ctx.getInputType(num_operands)->tensor_type()"? This may crash before we get to line 1714 if num_operands > numInputs, right?

Reply:
May be a test-case with num_operands > num_inputs would also help.

Reply:
Sure. Does it make sense to have a negative shape inference test?

Reply:
I think so. Test cases to check that the behavior is acceptable when preconditions are not satisfied are useful. E.g., something like in: https://github.com/onnx/onnx/blob/dfa4384c3f8fe9d804032f992e6f818fc54152b0/onnx/test/shape_inference_test.py#L109